tt_metal/impl/dispatch/kernels/vc_packet_router.cpp

```diff
@@ -0,0 +1,306 @@
+                        remote_rx_x[i], remote_rx_y[i], remote_rx_queue_id[i], remote_rx_network_type[i],
+                        input_packetize[i], input_packetize_log_page_size[i],
+                        input_packetize_local_sem[i], input_packetize_upstream_sem[i],
+                        input_packetize_src_endpoint[i], input_packetize_dest_endpoint[i]);
+    }
+
+    if (!wait_all_src_dest_ready(input_queues, router_lanes, output_queues, router_lanes, timeout_cycles)) {
+        write_test_results(test_results, PQ_TEST_STATUS_INDEX, PACKET_QUEUE_TEST_TIMEOUT);
+        return;
+    }
+
+    write_test_results(test_results, PQ_TEST_MISC_INDEX, 0xff000001);
+
+    uint32_t curr_input = 0;
+    bool timeout = false;
+    bool all_outputs_finished = false;
+    uint64_t data_words_sent = 0;
+    uint64_t iter = 0;
+    uint64_t start_timestamp = get_timestamp();
+    uint32_t progress_timestamp = start_timestamp & 0xFFFFFFFF;
+    uint32_t heartbeat = 0;
+    while (!all_outputs_finished && !timeout) {
+        IDLE_ERISC_HEARTBEAT_AND_RETURN(heartbeat);
+        iter++;
+        if (timeout_cycles > 0) {
+            uint32_t cycles_since_progress = get_timestamp_32b() - progress_timestamp;
+            if (cycles_since_progress > timeout_cycles) {
+                timeout = true;
+                break;
+            }
+        }
+        if (input_queues[curr_input].get_curr_packet_valid()) {
+            bool full_packet_sent;
+            uint32_t words_sent = output_queues[curr_input].forward_data_from_input(0, full_packet_sent, input_queues[curr_input].get_end_of_cmd());
+            data_words_sent += words_sent;
+            if ((words_sent > 0) && (timeout_cycles > 0)) {
+                progress_timestamp = get_timestamp_32b();
+            }
+        }
+
+        output_queues[curr_input].prev_words_in_flight_check_flush();
+
+        all_outputs_finished = true;
+        for (uint32_t i = 0; i < router_lanes; i++) {
+            all_outputs_finished &= output_queues[i].is_remote_finished();
+        }
+
+        curr_input++;
+        if (curr_input == router_lanes) {
+            curr_input = 0;
+        }
+    }
+
+    if (!timeout) {
+        write_test_results(test_results, PQ_TEST_MISC_INDEX, 0xff000002);
+        for (uint32_t i = 0; i < router_lanes; i++) {
+            if (!output_queues[i].output_barrier(timeout_cycles)) {
+                timeout = true;
+                break;
+            }
+        }
+    }
+
+    uint64_t cycles_elapsed = get_timestamp() - start_timestamp;
+    if (!timeout) {
+        write_test_results(test_results, PQ_TEST_MISC_INDEX, 0xff000003);
+        for (uint32_t i = 0; i < router_lanes; i++) {
+            input_queues[i].send_remote_finished_notification();
+        }
+    }
+
+    set_64b_result(test_results, data_words_sent, PQ_TEST_WORD_CNT_INDEX);
+    set_64b_result(test_results, cycles_elapsed, PQ_TEST_CYCLES_INDEX);
+    set_64b_result(test_results, iter, PQ_TEST_ITER_INDEX);
+
```

Comment:
NIT: the first 2 results we write are updated/computed on the critical path. Not high overhead, but do we need these in production mode?